crates/witness-generator/src/rpc.rs

```diff
@@ -0,0 +1,142 @@
+        self.header_map = headers
+            .into_iter()
+            .map(|(name, value)| {
+                Ok((
+                    HeaderName::from_str(name.as_ref())?,
+                    HeaderValue::from_str(value.as_ref())?,
+                ))
+            })
+            .collect::<Result<HeaderMap>>()?;
+        Ok(self)
+    }
+
+    /// Sets the numbe of last blocks to fetch.
+    pub fn last_n_blocks(mut self, n: usize) -> Self {
+        self.last_n_blocks = Some(n);
+        self
+    }
+
+    /// Builds the configured `RPCBlocksAndWitnesses`.
+    pub fn build(self) -> Result<RPCBlocksAndWitnesses> {
+        let client = HttpClientBuilder::default()
+            .set_headers(self.header_map)
+            .max_response_size(1 << 30)
+            .build(&self.url)?;
+        Ok(RPCBlocksAndWitnesses {
+            client,
+            last_n_blocks: self.last_n_blocks.unwrap_or(1),
+        })
+    }
+}
+
+/// RPCBlocksAndWitnesses is a witness generator that fetches blocks and witnesses
+#[derive(Debug, Clone)]
+pub struct RPCBlocksAndWitnesses {
+    client: HttpClient,
+    last_n_blocks: usize,
+}
+
+#[async_trait]
+impl WitnessGenerator for RPCBlocksAndWitnesses {
+    async fn generate(&self) -> Result<Vec<BlocksAndWitnesses>> {
+        if self.last_n_blocks == 0 {
+            return Ok(vec![]);
+        }
+
+        let latest_block = EthApiClient::<Transaction, Block, Receipt, Header>::block_by_number(
+            &self.client,
+            BlockNumberOrTag::Latest,
+            false,
+        )
+        .await?
+        .ok_or(anyhow::anyhow!("No block found"))?;
+
+        let (block_num_start, block_num_end) = (
+            max(
+                0,
+                latest_block.header.number - (self.last_n_blocks as u64 - 1),
+            ),
+            latest_block.header.number,
+        );
+
+        let mut hashes = Vec::with_capacity(self.last_n_blocks);
+        hashes.push((latest_block.header.number, latest_block.header.hash));
+        for n in (block_num_start..block_num_end).rev() {
+            let block_hash = hashes.last().unwrap().1;
+            let block = EthApiClient::<Transaction, Block, Receipt, Header>::block_by_hash(
+                &self.client,
+                block_hash,
+                true,
+            )
+            .await?
+            .ok_or(anyhow::anyhow!("No block found for number {}", n))?;
+            hashes.push((n, block.header.parent_hash));
+        }
+
+        let mut blocks_and_witnesses = Vec::with_capacity(hashes.len());
+        for (block_num, block_hash) in hashes {
+            let witness = self
+                .client
+                // TODO: our current reth RPC version doesn't support calling `debug_executionWitnessByHash`.
+                // Temporarily use the "by number" version which is potentially incorrect if a reorg happens.
+                .debug_execution_witness(BlockNumberOrTag::Number(block_num))
```

Comment:
I need to ask the Devops team to enable the `debug_executionWitnessFromHash` API since they have it blocked. 

That's the ideal API to call. This other one is "fine" but is a bit unsafe relying on block numbers since that could break the "blockchain consistency" if there's a reorg in the middle of the RPC calls.

Not a huge deal but just leaving the comment for full correctness until devops can enable this other API.